WordPress/pattern-directory
# Patterns: Update URL + query logic for consistent navigation

Fixes #258 by refactoring how the query object is generated from URLs/paths, and how URLs & context messages are generated from the query object. This collects logic around the app for getting filter view parameters into one place, which can convert to/from URLs & query objects.

Introduces `getQueryFromUrl` and `getUrlFromQuery`, so we're no longer getting the query values piecemeal with `getCategoryFromPath` etc. None of the `get*FromPath` functions should be used outside of the selector (the exception is the MyPatterns component). Also, we don't need to build up category paths anymore - we can use `getUrlFromQuery( { ...query, 'pattern-categories': id } ),` and it will build the correct URL, complete with current `orderby` or `author` properties.

Additionally, I refactored the `CategoryContextBar` to show the current author on author archives, plus sub-category navigation. It only shows the username for now, post-launch we can revisit to get the author's display name.

⚠️  I know this is a big change to how site nav works, and if it looks good I'd like to merge it ASAP - it would be great to have the author stuff working for launch. But let me know if anything doesn't look right, and I'll iterate.

### Screenshots

<img width="982" alt="Screen Shot 2021-07-19 at 7 30 51 PM" src="https://user-images.githubusercontent.com/541093/126240226-ebc90982-4de7-4fb1-9bbe-32c0c40b643d.png">

<img width="530" alt="Screen Shot 2021-07-19 at 7 20 07 PM" src="https://user-images.githubusercontent.com/541093/126240181-5a256811-3725-4fd5-affe-64960ca91d30.png">

<img width="579" alt="Screen Shot 2021-07-19 at 7 20 31 PM" src="https://user-images.githubusercontent.com/541093/126240183-c40f2682-96ca-4657-afab-5b65f43fdecd.png">

### How to test the changes in this Pull Request:

1. I added tests for the new query selectors, those should pass.
2. Try clicking around the site- the patterns should load correctly for all filter views you pick

Some example paths:
- Click on a category, view page 2 or 3, clicking on an author in the category view should bring you to all patterns by that author
- View an author archive, view page 2 or 3, then click on a cateogry - it should show you page 1 of that author's patterns in the chosen category.
